### ISSUE TITLE
Project challenges, insights and lessons learned

### ISSUE THREAD

**Issue by tenongene** [tenongene/quotegen_ecs#3]
- The with the initial phase of the project with EC2 instances, my user data script was not bootstrapping the EC2 instances with docker in order to pull images and then run the container:
Further research pointed out that there should be no indentation to the right of the user data script, in the terraform manifest file. When I took out the indentation, the EC2 instances then bootstrapped correctly with docker running.
I also learned, only after completing the 2nd phase of the project with ECS that I also had the option of using an Amazon Linux ECS optimized AMI, which already comes with docker pre-installed.

![Image](https://user-images.githubusercontent.com/49034904/263775911-c94bb652-97c0-4611-94ba-93b426ab0d15.png)

Indentations in user data script

![Image](https://user-images.githubusercontent.com/49034904/263776119-fe10ce48-86c2-494c-84e8-bea9fe77dd0a.png)

Do not indent.

====================================================================

- I initially could not reach the application although everything seemed to have been configured correctly.
My debugging method was to access the instances via SSH and check the docker logs. I noticed that docker could not load the express module in the application during the image pull step. That means I forgot to run "NPM install" to install the dependencies from the package.json during image the building step. I added this, and rebuilt the image and then the image pulled correctly.

![Image](https://user-images.githubusercontent.com/49034904/263774175-e7dee8f1-47d6-4ae4-b356-f098db1f3d70.png)

![Image](https://user-images.githubusercontent.com/49034904/263774200-d66bc2dc-d8e5-4677-bdcb-869d1ed9a42d.png)

=======================================================================

- During the ECS deployment phase of the project, at first, although things seemed to be configured correctly, the tasks could not start and none of the containers could be deployed for some reason. I couldn't find any issues and the EC2 instances where running and passing system checks.
- Upon further research, I learned I could check the logs of the ECS agent at /var/log/ecs/ecs-agent.log to get a clue of what could be wrong. I checked the logs and saw that the ECS agent did not have the proper credentials and permissions to pull images. My understanding was that an instance profile was required to do so. So I created an instance profile and attached to the ASG definition and that got resolved.

![Image](https://user-images.githubusercontent.com/49034904/263806998-b1d880e7-d32f-448c-8a73-56e8f14db86e.png)

![Image](https://user-images.githubusercontent.com/49034904/263807064-2b1f249f-f17b-4af0-b9fd-b0bb0646eee9.png)

=======================================================================

After that resolution the new issue was an error that the containers could not be registered because the cluster was inactive, although I could clearly see on the console that the cluster was active.
Further research with an answer from stack overflow suggested that the ECS agent only creates a "default" cluster, unless the cluster is specifically added as an environment  variable (ECS_CLUSTER) and added to the ECS agent's configuration. I added it via a user data script and everything got resolved.

![Image](https://user-images.githubusercontent.com/49034904/263808840-9d6365de-7de5-493b-9dd0-79f469f2f8ca.png)

![Image](https://user-images.githubusercontent.com/49034904/263808951-46230f7c-844f-4098-a7cf-cf051ff59bdd.png)

![Image](https://user-images.githubusercontent.com/49034904/263809055-5b0008c0-569d-4a10-85c2-3a8150a7b675.png)

![Image](https://user-images.githubusercontent.com/49034904/263809234-903c405e-660f-4dd2-afbb-f79277e896ec.png)

========================================================================

After the first phase of the project was completed, I noticed that the lambda function I had written was not optimized to reduce execution time and thus cloud costs, as recommended by lambda best practices. I defined several constants within the handler, which increases execution time. 
With the ECS version of the function, I ensured to declare all constants outside the handler, and the result was almost a 50% reduction in execution time.

![Image](https://github.com/tenongene/quotegen_ecs/assets/49034904/5d77fb12-c938-45c6-961f-ea8503ac067f)

Constants declared within handler....not ideal.

========

![Image](https://github.com/tenongene/quotegen_ecs/assets/49034904/d500a523-e100-400e-98db-2c17813ff07c)

Initially did the same for the ECS lambda function, but then recognized this and refactored the code by moving all constants declared outside the function handler.

=========

![Image](https://github.com/tenongene/quotegen_ecs/assets/49034904/e86c4578-b892-4425-91a2-7a30a1e9fb6d)

Moving constants outside the lambda function handler improved the execution time from 1447ms to 740ms.
